solana/programs/ntt-quoter/src/processor/request_relay.rs

```diff
@@ -0,0 +1,138 @@
+    #[account(
+    init,
+    payer = payer,
+    space = 8 + RelayRequest::INIT_SPACE,
+    seeds = [RelayRequest::SEED_PREFIX, outbox_item.key().as_ref()],
+    bump,
+  )]
+    pub relay_request: Account<'info, RelayRequest>,
+
+    pub system_program: Program<'info, System>,
+}
+
+#[derive(AnchorSerialize, AnchorDeserialize)]
+pub struct RequestRelayArgs {
+    pub gas_dropoff: u64, //NativeAmount,
+    pub max_fee: u64,     //SolAmount,
+}
+
+const GWEI: u64 = u64::pow(10, 9);
+
+fn mul_div(scalar: u64, numerator: u64, denominator: u64) -> u64 {
+    ((scalar as u128) * (numerator as u128) / (denominator as u128))
+        .try_into()
+        .unwrap()
+}
+
+//TODO built-in u128 division likely still wastes a ton of compute units
+//     might be more efficient to use f64 or ruint crate
+pub fn request_relay(ctx: Context<RequestRelay>, args: RequestRelayArgs) -> Result<()> {
+    let accs = ctx.accounts;
+
+    require_gt!(
+        args.gas_dropoff,
+        accs.registered_chain.max_gas_dropoff,
+        NttQuoterError::ExceedsMaxGasDropoff
+    );
+
+    let total_fee_in_lamports_without_rent = {
+        let target_native_in_gwei = args.gas_dropoff
+            + if accs.registered_chain.gas_price > 0 {
+                //gas_price[wei, 18 decimals] * EVM_GAS_COST / GWEI = target_native[gwei, 9 decimals]
+                mul_div(accs.registered_chain.gas_price, EVM_GAS_COST, GWEI)
+            } else {
+                0
+            };
+
+        //usd/target_native[usd, 6 decimals] * target_native[gwei, 9 decimals] = usd[usd, 6 decimals]
+        let target_native_in_usd = mul_div(
+            accs.registered_chain.native_price,
+            target_native_in_gwei,
+            GWEI,
+        );
+
+        let total_in_usd = target_native_in_usd + accs.registered_chain.base_price;
+
+        if total_in_usd > 0 {
+            //total_fee[sol, 9 decimals] = total_usd[usd, 6 decimals] / (sol_price[usd, 6 decimals]
+            mul_div(total_in_usd, LAMPORTS_PER_SOL, accs.instance.sol_price)
+        } else {
+            0
+        }
+    };
+
+    let rent_in_lamports = sysvar::rent::Rent::get()?.minimum_balance(8 + RelayRequest::INIT_SPACE);
+    let total_fee_in_lamports = total_fee_in_lamports_without_rent + rent_in_lamports;
+
+    require_gte!(
+        args.max_fee,
+        total_fee_in_lamports,
+        NttQuoterError::ExceedsUserMaxFee
+    );
+
+    msg!("total fee in lamports: {}", total_fee_in_lamports);
+
+    //store the requested gas dropoff
+    accs.relay_request.requested_gas_dropoff = args.gas_dropoff;
+
+    //pay the relayer
+    if total_fee_in_lamports_without_rent > 0 {
+        system_program::transfer(
+            CpiContext::new(
+                accs.system_program.to_account_info(),
+                Transfer {
+                    from: accs.payer.to_account_info(),
+                    to: accs.fee_recipient.to_account_info(),
```

Comment:
we could consider a simplification here: instead of transferring to the `fee_recipient`, just transfer to the `relay_request` account. Then the `fee_recipient` doesn't even need to be stored in the instance, and can just passed as an argument (well, an account) when closing the `relay_request` account. It also allows us to better see how much payment was made for an individual transfer by just looking at the account rather than logs and whatnot.

Reply:
That was the original plan, but then I figured it would actually be operationally easier to have the lion's share of the fee flow back to the fee recipient immediately without having to rely on running some off-chain process to collect them, which might be helpful during times of high through-put. Otoh it also introduces account competition around the fee recipient, which sucks so perhaps having the fully parallel, more legible solution is cleaner.